File: infinity-ledger-main/MEF-Core_v1.0/src/bench/drivers/qdrant_driver.py
"""Driver for Qdrant's HTTP API."""

from __future__ import annotations

import os
from typing import Iterable, List, Optional, Sequence, Tuple

from qdrant_client import QdrantClient
from qdrant_client.http import models as rest_models
from qdrant_client.http.exceptions import UnexpectedResponse

from .base import DriverUnavailable, UpsertItem, Vector, VectorStoreDriver


class QdrantDriver(VectorStoreDriver):
    """Benchmark driver that interacts with a Qdrant deployment via qdrant-client."""

    name = "Qdrant"

    def __init__(self, metric: str = "cosine") -> None:
        super().__init__(metric)
        self._base_url = (os.getenv("QDRANT_URL") or "").rstrip("/")
        self._client: Optional[QdrantClient] = None
        self._dimension: Optional[int] = None

    def connect(self) -> None:
        if not self._base_url:
            raise DriverUnavailable(self.name, "QDRANT_URL not configured")
        try:
            client = QdrantClient(url=self._base_url, prefer_grpc=False, timeout=5.0)
            client.get_health()
        except Exception as exc:  # pragma: no cover - network flakes
            raise DriverUnavailable(self.name, f"unable to connect to Qdrant: {exc}") from exc
        self._client = client

    def clear(self, namespace: str) -> None:
        client = self._require_client()
        try:
            client.delete_collection(namespace)
        except UnexpectedResponse as exc:
            if exc.status_code != 404:
                raise
        except Exception as exc:  # pragma: no cover - defensive
            raise DriverUnavailable(self.name, f"failed to clear collection: {exc}") from exc
        self._dimension = None

    def upsert(
        self,
        items: Iterable[UpsertItem],
        namespace: str,
        batch_size: int = 1000,
    ) -> None:
        client = self._require_client()
        batch: List[rest_models.PointStruct] = []
        for identifier, vector, metadata in items:
            if self._dimension is None:
                self._dimension = len(vector)
                self._ensure_collection(client, namespace, self._dimension)
            payload = rest_models.PointStruct(
                id=identifier,
                vector=[float(value) for value in vector],
                payload=dict(metadata) if metadata else None,
            )
            batch.append(payload)
            if len(batch) >= batch_size:
                self._flush(client, namespace, batch)
                batch = []
        if batch:
            self._flush(client, namespace, batch)

    def search(self, query: Vector, k: int, namespace: str) -> Sequence[Tuple[str, float]]:
        client = self._require_client()
        try:
            results = client.search(
                collection_name=namespace,
                query_vector=[float(value) for value in query],
                limit=int(k),
                with_payload=False,
                with_vectors=False,
            )
        except Exception as exc:  # pragma: no cover - defensive
            raise DriverUnavailable(self.name, f"search failed: {exc}") from exc
        hits: List[Tuple[str, float]] = []
        for entry in results:
            identifier = entry.id
            score = entry.score if entry.score is not None else 0.0
            hits.append((str(identifier), float(score)))
        return hits[:k]

    def _require_client(self) -> QdrantClient:
        if self._client is None:
            raise RuntimeError("connect() must be called before using the driver")
        return self._client

    def _ensure_collection(self, client: QdrantClient, namespace: str, dimension: int) -> None:
        distance = {
            "cosine": rest_models.Distance.COSINE,
            "ip": rest_models.Distance.DOT,
            "l2": rest_models.Distance.EUCLID,
        }.get(self.metric, rest_models.Distance.COSINE)
        vectors_config = rest_models.VectorParams(size=int(dimension), distance=distance)
        try:
            client.recreate_collection(
                collection_name=namespace,
                vectors_config=vectors_config,
                wait=True,
            )
        except Exception as exc:  # pragma: no cover - defensive
            raise DriverUnavailable(self.name, f"failed to ensure collection: {exc}") from exc

    def _flush(
        self,
        client: QdrantClient,
        namespace: str,
        batch: List[rest_models.PointStruct],
    ) -> None:
        try:
            client.upsert(collection_name=namespace, points=batch, wait=True)
        except Exception as exc:  # pragma: no cover - defensive
            raise DriverUnavailable(self.name, f"failed to upsert batch: {exc}") from exc


__all__ = ["QdrantDriver"]

File: mef-bench/src/qdrant_driver.rs
/*!
 * Driver for Qdrant's HTTP API.
 * 
 * Migrated from MEF-Core_v1.0/src/bench/drivers/qdrant_driver.py
 * 
 * This implementation uses the Qdrant HTTP API directly via reqwest
 * instead of the qdrant-client library to minimize dependencies.
 */

use crate::base::{DriverUnavailable, UpsertItem, Vector, VectorStoreDriver};
use anyhow::{Context, Result};
use serde_json::json;

/// Benchmark driver that interacts with a Qdrant deployment via HTTP API
pub struct QdrantDriver {
    metric: String,
    base_url: String,
    client: Option<reqwest::blocking::Client>,
    dimension: Option<usize>,
}

impl QdrantDriver {
    /// Create a new Qdrant driver
    pub fn new(metric: Option<&str>) -> Self {
        let metric = metric.unwrap_or("cosine").to_lowercase();
        let base_url = std::env::var("QDRANT_URL")
            .unwrap_or_default()
            .trim_end_matches('/')
            .to_string();
        
        Self {
            metric,
            base_url,
            client: None,
            dimension: None,
        }
    }

    fn ensure_collection(&mut self, namespace: &str, dimension: usize) -> Result<()> {
        let client = self.client.as_ref()
            .ok_or_else(|| anyhow::anyhow!("connect() must be called before using the driver"))?;

        let distance = match self.metric.as_str() {
            "cosine" => "Cosine",
            "ip" => "Dot",
            "l2" => "Euclid",
            _ => "Cosine",
        };

        let payload = json!({
            "vectors": {
                "size": dimension,
                "distance": distance
            }
        });

        let url = format!("{}/collections/{}", self.base_url, namespace);
        
        // Try to recreate the collection
        let _ = client
            .delete(&url)
            .timeout(std::time::Duration::from_secs(15))
            .send();

        let response = client
            .put(&url)
            .json(&payload)
            .timeout(std::time::Duration::from_secs(15))
            .send()
            .map_err(|e| {
                DriverUnavailable::new(
                    "Qdrant",
                    format!("failed to ensure collection: {}", e),
                )
            })?;

        if !response.status().is_success() {
            return Err(DriverUnavailable::new(
                "Qdrant",
                format!("failed to create collection: HTTP {}", response.status()),
            )
            .into());
        }

        Ok(())
    }

    fn flush_batch(&self, namespace: &str, batch: Vec<serde_json::Value>) -> Result<()> {
        let client = self.client.as_ref()
            .ok_or_else(|| anyhow::anyhow!("connect() must be called before using the driver"))?;

        let payload = json!({
            "points": batch,
            "wait": true
        });

        let url = format!("{}/collections/{}/points", self.base_url, namespace);
        let response = client
            .put(&url)
            .json(&payload)
            .timeout(std::time::Duration::from_secs(30))
            .send()
            .map_err(|e| {
                DriverUnavailable::new(
                    "Qdrant",
                    format!("failed to upsert batch: {}", e),
                )
            })?;

        if !response.status().is_success() {
            return Err(DriverUnavailable::new(
                "Qdrant",
                format!("failed to upsert batch: HTTP {}", response.status()),
            )
            .into());
        }

        Ok(())
    }
}

impl VectorStoreDriver for QdrantDriver {
    fn name(&self) -> &str {
        "Qdrant"
    }

    fn metric(&self) -> &str {
        &self.metric
    }

    fn connect(&mut self) -> Result<()> {
        if self.base_url.is_empty() {
            return Err(DriverUnavailable::new(
                "Qdrant",
                "QDRANT_URL not configured".to_string(),
            )
            .into());
        }

        let client = reqwest::blocking::Client::builder()
            .timeout(std::time::Duration::from_secs(30))
            .build().context("Failed to build HTTP client")?;

        let health_url = format!("{}/health", self.base_url);
        let response = client
            .get(&health_url)
            .timeout(std::time::Duration::from_secs(5))
            .send()
            .map_err(|e| {
                DriverUnavailable::new(
                    "Qdrant",
                    format!("unable to connect to Qdrant: {}", e),
                )
            })?;

        if !response.status().is_success() {
            return Err(DriverUnavailable::new(
                "Qdrant",
                format!("health check failed: HTTP {}", response.status()),
            )
            .into());
        }

        self.client = Some(client);
        Ok(())
    }

    fn clear(&mut self, namespace: &str) -> Result<()> {
        let client = self.client.as_ref()
            .ok_or_else(|| anyhow::anyhow!("connect() must be called before using the driver"))?;

        let url = format!("{}/collections/{}", self.base_url, namespace);
        let response = client
            .delete(&url)
            .timeout(std::time::Duration::from_secs(15))
            .send()
            .map_err(|e| {
                DriverUnavailable::new(
                    "Qdrant",
                    format!("failed to clear collection: {}", e),
                )
            })?;

        let status = response.status().as_u16();
        if status != 200 && status != 404 {
            return Err(DriverUnavailable::new(
                "Qdrant",
                format!("failed to clear collection: HTTP {}", status),
            )
            .into());
        }

        self.dimension = None;
        Ok(())
    }

    fn upsert(
        &mut self,
        items: Vec<UpsertItem>,
        namespace: &str,
        batch_size: usize,
    ) -> Result<()> {
        if self.client.is_none() {
            return Err(anyhow::anyhow!("connect() must be called before using the driver"));
        }

        let mut batch: Vec<serde_json::Value> = Vec::new();

        for (identifier, vector, metadata) in items {
            if self.dimension.is_none() {
                self.dimension = Some(vector.len());
                self.ensure_collection(namespace, vector.len())?;
            }

            let point = json!({
                "id": identifier,
                "vector": vector,
                "payload": metadata.unwrap_or_default()
            });

            batch.push(point);

            if batch.len() >= batch_size {
                self.flush_batch(namespace, batch.clone())?;
                batch.clear();
            }
        }

        if !batch.is_empty() {
            self.flush_batch(namespace, batch)?;
        }

        Ok(())
    }

    fn search(
        &self,
        query: &Vector,
        k: usize,
        namespace: &str,
    ) -> Result<Vec<(String, f64)>> {
        let client = self.client.as_ref()
            .ok_or_else(|| anyhow::anyhow!("connect() must be called before using the driver"))?;

        let search_body = json!({
            "vector": query,
            "limit": k,
            "with_payload": false,
            "with_vector": false
        });

        let url = format!("{}/collections/{}/points/search", self.base_url, namespace);
        let response = client
            .post(&url)
            .json(&search_body)
            .timeout(std::time::Duration::from_secs(15))
            .send()
            .map_err(|e| {
                DriverUnavailable::new(
                    "Qdrant",
                    format!("search failed: {}", e),
                )
            })?;

        if !response.status().is_success() {
            return Err(DriverUnavailable::new(
                "Qdrant",
                format!("search failed: HTTP {}", response.status()),
            )
            .into());
        }

        let payload: serde_json::Value = response.json().context("Failed to parse search response")?;

        let results = payload
            .get("result")
            .and_then(|r| r.as_array())
            .ok_or_else(|| anyhow::anyhow!("Invalid search response format"))?;

        let mut hits: Vec<(String, f64)> = Vec::new();
        for entry in results {
            if let (Some(id), Some(score)) = (
                entry.get("id").and_then(|v| v.as_str().or_else(|| v.as_u64().map(|n| Box::leak(n.to_string().into_boxed_str()) as &str))),
                entry.get("score").and_then(|v| v.as_f64()),
            ) {
                hits.push((id.to_string(), score));
            }
        }

        Ok(hits.into_iter().take(k).collect())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_qdrant_driver_creation() {
        let driver = QdrantDriver::new(None);
        assert_eq!(driver.name(), "Qdrant");
        assert_eq!(driver.metric(), "cosine");
    }

    #[test]
    fn test_qdrant_driver_custom_metric() {
        let driver = QdrantDriver::new(Some("l2"));
        assert_eq!(driver.metric(), "l2");
    }

    #[test]
    fn test_qdrant_driver_ip_metric() {
        let driver = QdrantDriver::new(Some("ip"));
        assert_eq!(driver.metric(), "ip");
    }

    #[test]
    fn test_qdrant_driver_base_url_env() {
        std::env::set_var("QDRANT_URL", "http://localhost:6333/");
        let driver = QdrantDriver::new(None);
        assert_eq!(driver.base_url, "http://localhost:6333");
        std::env::remove_var("QDRANT_URL");
    }

    #[test]
    fn test_qdrant_driver_default_url() {
        std::env::remove_var("QDRANT_URL");
        let driver = QdrantDriver::new(None);
        assert_eq!(driver.base_url, "");
    }

    #[test]
    fn test_connect_without_url() {
        let mut driver = QdrantDriver::new(None);
        let result = driver.connect();
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("QDRANT_URL not configured"));
    }

    #[test]
    fn test_upsert_without_connect() {
        let mut driver = QdrantDriver::new(Some("cosine"));
        let items = vec![
            ("id1".to_string(), vec![1.0, 2.0, 3.0], None),
        ];
        let result = driver.upsert(items, "test", 1000);
        assert!(result.is_err());
    }

    #[test]
    fn test_search_without_connect() {
        let driver = QdrantDriver::new(Some("cosine"));
        let query = vec![1.0, 2.0, 3.0];
        let result = driver.search(&query, 10, "test");
        assert!(result.is_err());
    }

    #[test]
    fn test_clear_without_connect() {
        let mut driver = QdrantDriver::new(Some("cosine"));
        let result = driver.clear("test");
        assert!(result.is_err());
    }
}
